core/src/main/java/org/elasticsearch/search/aggregations/support/SingleValuesSourceAggregatorBuilder.java

```diff
@@ -0,0 +1,216 @@
+
+    public static abstract class LeafOnly<VS extends ValuesSource, AB extends SingleValuesSourceAggregatorBuilder<VS, AB>>
+        extends SingleValuesSourceAggregatorBuilder<VS, AB> {
+
+        protected LeafOnly(String name, Type type, ValuesSourceType valuesSourceType, ValueType targetValueType) {
+            super(name, type, valuesSourceType, targetValueType);
+        }
+
+        /**
+         * Read an aggregation from a stream that does not serialize its targetValueType. This should be used by most subclasses.
+         */
+        protected LeafOnly(StreamInput in, Type type, ValuesSourceType valuesSourceType, ValueType targetValueType) throws IOException {
+            super(in, type, valuesSourceType, targetValueType);
+        }
+
+        /**
+         * Read an aggregation from a stream that serializes its targetValueType. This should only be used by subclasses that override
+         * {@link #serializeTargetValueType()} to return true.
+         */
+        protected LeafOnly(StreamInput in, Type type, ValuesSourceType valuesSourceType) throws IOException {
+            super(in, type, valuesSourceType);
+        }
+
+        @Override
+        public AB subAggregations(Builder subFactories) {
+            throw new AggregationInitializationException("Aggregator [" + name + "] of type [" + type + "] cannot accept sub-aggregations");
+        }
+    }
+
+    private String field;
+    private Script script;
+
+    protected SingleValuesSourceAggregatorBuilder(String name, Type type, ValuesSourceType valuesSourceType, ValueType targetValueType) {
+        super(name, type, valuesSourceType, targetValueType);
+    }
+
+    /**
+     * Read an aggregation from a stream that does not serialize its targetValueType. This should be used by most subclasses.
+     */
+    protected SingleValuesSourceAggregatorBuilder(StreamInput in, Type type, ValuesSourceType valuesSourceType, ValueType targetValueType)
+        throws IOException {
+        super(in, type, valuesSourceType, targetValueType);
+    }
+
+    /**
+     * Read an aggregation from a stream that serializes its targetValueType. This should only be used by subclasses that override
+     * {@link #serializeTargetValueType()} to return true.
+     */
+    protected SingleValuesSourceAggregatorBuilder(StreamInput in, Type type, ValuesSourceType valuesSourceType) throws IOException {
+        super(in, type, valuesSourceType);
+    }
+
+    /**
+     * Read from a stream.
+     */
+    protected void read(StreamInput in) throws IOException {
+        field = in.readOptionalString();
+        if (in.readBoolean()) {
+            script = new Script(in);
+        }
+        if (in.readBoolean()) {
+            valueType = ValueType.readFromStream(in);
+        }
+        format = in.readOptionalString();
+        missing = in.readGenericValue();
+        if (in.readBoolean()) {
+            timeZone = DateTimeZone.forID(in.readString());
```

Comment:
Really the attributes for the super class should be read and written by the superclass to ensure they are consistent across all implementations